KEP-3766: Move ReferenceGrant to sig-auth API Group (#3767)
kubernetes/enhancements: keps/sig-auth/3766-referencegrant/README.md

_this_ API requires the additional specificity that `Resource` provides.

The Gateway API ReferenceGrant looked like this:
```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: ReferenceGrant
metadata:
  name: allow-gateways
  namespace: bar
spec:
  from:
    # Note that in Gateway API, Group is currently defaulted
    # to this, which means you to explicitly set the group to
    # the empty string for Core resources. We should definitely
    # change this.
    - group: "gateway.networking.kubernetes.io"
      kind: Gateway
      namespace: foo
  to:
   - group: ""
     kind: Secret
```

The new version will look like this instead:
```yaml
apiVersion: authorization.k8s.io/v1alpha1
kind: ReferenceGrant
metadata:
  name: allow-gateways
  namespace: bar
spec:
  from:
    # Assuming that we leave the default for Group to the empty
    # string, so that Core objects don't need additional config.
    - group: "gateway.networking.kubernetes.io" 
      resource: gateways
      namespace: foo
  to:
    - resource: secrets

```

The new version communicates the scope more clearly because `resource` is
unambiguous and corresponds to exactly one URL path on the API Server.

> 

Reviewer: ```suggestion

This change also leaves room for an enhancement.
Whether we have an in-tree or CRD implementation, we can rely on the exact matching that
the plural resource name gives us, and
[warn](https://kubernetes.io/blog/2020/09/03/warnings/) if either side of the grant is for
an API that's not served by this cluster.
```

Depending on what the eventual scope of issue 3766 becomes, we might  one day also add matching on
`uid` (in case the specific UID matters).

Author: Updated, thanks!

But in this case, the cross-namespace reference is the _only_ backend, so the
removal of the ReferenceGrant will also result in the removal of the HTTPRoute's
config from the data plane.

#### PersistentVolumeClaim using cross namespace data source

This example is taken from https://kubernetes.io/blog/2023/01/02/cross-namespace-data-sources-alpha/
and updated to use the proposed new spec.

It allows the PersistentVolumeClaim in the `dev` namespace to use a volume
snapshot from the `prod` namespace as its data source.

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: example-pvc
  namespace: dev
spec:
  storageClassName: example
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
  dataSourceRef:
    apiGroup: snapshot.storage.k8s.io
    kind: VolumeSnapshot
    name: new-snapshot-demo
    namespace: prod
  volumeMode: Filesystem
---
apiVersion: authorization.k8s.io/v1alpha1
kind: ReferenceGrant
metadata:
  name: allow-prod-pvc
  namespace: prod
spec:
  from:
  - resource: persistentvolumeclaims
    namespace: dev
  to:
  - group: snapshot.storage.k8s.io
    resource: volumesnapshots
    name: new-snapshot-demo
```

> When a ReferenceGrant is deleted, any existing snapshots enabled by this

Reviewer 2: small correction: "any existing volumes created from the cross-namespace datasource will still persist, but new volumes will be rejected"

Author: Updated, thanks!

conformance tests for their use of ReferenceGrant.

### Graduation Criteria

#### GA

[ ] Almost all of the fields and behavior have conformance test coverage.
[ ] Multiple conformant implementations.
[ ] Widespread implementation and usage.
[ ] Conformance tests that exercise all ReferenceGrant API calls (not the actual implementation of the API).
[ ] At least 6 months of soak time as a beta API (graduated to beta in Dec 2022).

### Upgrade / Downgrade Strategy

N/A

### Version Skew Strategy

Version skew is a bit different here since it won't be implemented in-tree.
There will be some implementation that support both the API defined by Gateway
API and this API. Since these resources are entirely additive and can be
duplicative, we can copy Gateway API resources to the new API group and delete
the old Gateway API resources as part of a seamless migration. We expect that
many implementations will provide this recommendation to users, and we may even
provide tooling to simplify this process.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Other

>   - Describe the mechanism: Enable beta ReferenceGrant API

Reviewer 3: ```suggestion
  - Describe the mechanism: Enable alpha ReferenceGrant API
```

Author: Updated, thanks!

ended up making https://github.com/kubernetes/community/pull/5973 to clarify
the API conventions.

However, in discussion on this KEP, it's clear that the more generic nature of
_this_ API requires the additional specificity that `Resource` provides.

The Gateway API ReferenceGrant looked like this:
```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: ReferenceGrant
metadata:
  name: allow-gateways
  namespace: bar
spec:
  from:
    # Note that in Gateway API, Group is currently defaulted
    # to this, which means you to explicitly set the group to
    # the empty string for Core resources. We should definitely
    # change this.
    - group: "gateway.networking.kubernetes.io"
      kind: Gateway
      namespace: foo
  to:
   - group: ""
     kind: Secret
```

The new version will look like this instead:
```yaml
apiVersion: authorization.k8s.io/v1alpha1
kind: ReferenceGrant
metadata:
  name: allow-gateways
  namespace: bar
spec:
  from:
    # Assuming that we leave the default for Group to the empty
    # string, so that Core objects don't need additional config.
    - group: "gateway.networking.kubernetes.io" 
      resource: gateways
      namespace: foo
  to:

>     - resource: secrets

Author: Updated, thanks!

```

> The new version communicates the scope more clearly because `resource` is
> unambiguous and corresponds to exactly one URL path on the API Server.

Reviewer 4: ```suggestion
The new version communicates the scope more clearly because `group`+`resource` is
unambiguous and corresponds to exactly one set of objects on the API Server.
```

Author: Updated, thanks!

is removed as data points.

### Example Usage

#### Gateway API Gateway Referencing Secret

In this example (from the Gateway API docs), we have a Gateway in the
`gateway-api-example-ns1` namespace, referencing a Secret in the
`gateway-api-example-ns2` namespace. The following ReferenceGrant allows this:

```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: cross-namespace-tls-gateway
  namespace: gateway-api-example-ns1
spec:
  gatewayClassName: acme-lb
  listeners:
  - name: https
    protocol: HTTPS
    port: 443
    hostname: "*.example.com"
    tls:
      # There's a Kind/Resource mismatch here, which sucks, but it is not
      # easily fixable, since Gateway is already a beta, close to GA
      # object.
      certificateRefs:
      - kind: Secret
        group: ""
        name: wildcard-example-com-cert
        namespace: gateway-api-example-ns2
---
apiVersion: authorization.k8s.io/v1alpha1
kind: ReferenceGrant
metadata:
  name: allow-ns1-gateways-to-ref-secrets
  namespace: gateway-api-example-ns2
spec:
  from:
  - group: gateway.networking.k8s.io
    resource: gateways
    namespace: gateway-api-example-ns1
  to:

>   - resource: secrets

Reviewer 4: let's err on the side of keeping group explicit in the examples for now

```suggestion
  - group: ""
    resource: secrets
```

Author: Updated, thanks!

```

For Gateway TLS references, if this ReferenceGrant is deleted (revoking, 
the grant), then the Listener will become invalid, and the configuration
will be removed as soon as possible (eventual consistency permitting).

#### Gateway API HTTPRoute Referencing Service

In this example, a HTTPRoute in the `baz` namespace is directing traffic
to a Service backend in the `quux` namespace.

```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: quuxapp
  namespace: baz
spec:
  parentRefs:
  - name: example-gateway
    sectionName: https
  hostnames:
  - quux.example.com
  rules:
  - matches:
    - path:
        type: PathPrefix
        value: /
    # BackendRefs are Services by default.
    backendRefs:
    - name: quuxapp
      namespace: quux
      port: 80
---
apiVersion: authorization.k8s.io/v1alpha1
kind: ReferenceGrant
metadata:
  name: allow-baz-httproutes
  namespace: quux
spec:
  from:
  - group: gateway.networking.k8s.io
    resource: httproutes
    namespace: baz
  to:

>   - resource: services

Reviewer 4: ```suggestion
  - group: ""
    resource: services
```

Author: Updated, thanks!

also strongly recommend that every API that relies on ReferenceGrant 
includes robust conformance tests covering this functionality. Existing
Gateway API conformance tests can serve as a model for this.

#### Cross-Namespace References may Weaken Namespace Boundaries
Although we believe that the handshake required for cross-namespace references
with ReferenceGrant ensures these references will be safe, it does potentially
weaken existing namespace boundaries. We believe ReferenceGrant will have a net
benefit on the ecosystem as it will allow workloads, secrets, and configuration
to be deployed in separate namespaces that more clearly match up with desired
authorization.

## Design Details

### General Notes

#### `ReferenceGrant` is half of a handshake

When thinking about ReferenceGrant, it is important to remember that it does not
do anything by itself. It *Grants* the *possibility* of making a *Reference*
across namespaces. It's intended that _another object_ (that is, the From object)
complete the handshake by creating a reference to the referent object (the To
object).

#### ReferenceGrant authors must have sufficient access

> Anyone creating or updating a ReferenceGrant MUST have read access to the resources

Reviewer 4: add an unresolved block (linking to https://github.com/kubernetes/enhancements/pull/3767#discussion_r1084528657 as context) about whether checks beyond read are needed (if not all grants are limited to granting read access)

Author: Updated, thanks!

### Do We Need Verbs?

Previous Discussion: https://github.com/kubernetes/enhancements/pull/3767#discussion_r1084509958

We could add a `verbs` field to enable users to specify the kind of referential
access they want to grant. For example, we could define "read", "route", and
"backup" as well-known verbs to start. We could also allow implementations to
support additional domain-prefixed verbs.

This would enable more precise grants, and potentially more open ended fields
elsewhere in the resource, see the next item for more.

### Do We Need Any or Wildcard Selectors?

Previous Discussion: https://github.com/kubernetes/enhancements/pull/3767#discussion_r1086020464

We already allow "Name" to be optional in `To`, effectively resulting in
wildcard behavior. Should we expand that to allow any of the following?

1. References to any group or resource
1. References from any group or resource
1. References from any namespace

Historically, there has been pushback to allowing any group or resource
because it could enable unknown or unintended kinds of access. If we
introduced the concepts of "verbs" described above, this would become
a moot point.

### Do We Need Label Selectors?

> Previous Discussion: https://github.com/kubernetes/enhancements/pull/3767#discussion_r1084492070

Reviewer 4: https://github.com/kubernetes/enhancements/pull/3767#discussion_r1084674648 was another thread about selectors you could link to as well

Author: Updated, thanks!

label selectors to enable:

1. Access to resources with specific labels
1. Access from namespaces with specific labels

In both cases, this would significantly increase implementation
complexity.

As a potential middleground, we could explore a solution that left
room for namespace selectors without actually including them. For example: 

```go
type ReferenceGrantFrom struct {
  //...
  Peer ReferenceGrantPeer
}

// Exatcly one field should be set. If none are found, clients must
// assume that an unknown value was specified.
type ReferenceGrantPeer {
    Namespace *string
    // Future: Namespace selector
}
```

This assumes that group+resource supersedes namespaces - is that true? Or do we
really want:

```go
type ReferenceGrant struct {
    // ...
    From ReferenceGrantPeer
}

type ReferenceGrantPeer struct {
    // ...
    Namespace *string
    NamespaceResource *NamespaceResource
}

type NamespaceResource struct {
    Namespace string
    Group string
    Resource string
}
```
<<[/UNRESOLVED]>>

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Other

>   - Describe the mechanism: Enable alpha ReferenceGrant API

Reviewer 5: How is this done, is there a feature gate for this?

  - Will enabling / disabling the feature require downtime of the control
    plane?
    No
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
    No

###### Does enabling the feature change any default behavior?

No

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

> Yes

Reviewer 5: How is this done? Would controllers using this API need to be removed?
